### PR TITLE
Support throw error with variables

### LINKED ISSUE #4080: Update variables on BPMN error
At the moment you can't implement a retry loop with a timer that is entered via a BPMN error throw and has decrementing retries:

Screen Shot 2020-03-20 at 1 24 26 am

What I'd like to be able to do is this: 

Screen Shot 2020-03-20 at 1 25 52 am

However, there is no opportunity to mutate anything when throwing a bpmn error from the client....

### LINKED ISSUE #4337: Save/extract error code, error message to variables in Error Catch Boundary Event
**Is your feature request related to a problem? Please describe.**
Currently when error is caught, errorCode and errorMessage are discarded and cannot be saved to variable.

**Describe the solution you'd like**
In Error Catch Boundary Event, save errorCode and errorMessage (and desirably, a custom object similar to regular worker results) to variable, probably similar to/improve [how Camunda/Camunda Modeler implemented it](https://forum.camunda.org/t/error-message-of-bpmn-error-from-external-task/5519/5?u=ceefour):

image

I myself think that error code+message+variables should be one object like it's usual in the REST/Spring Boot REST world (however this will drift from Camunda as Camunda treats errorCode and errorMessage separately, and does not have error results variables):

```json
{
 "error": "com.mycompany.http.BadRequestException",
 "message": "JSON parse error: Unrecognized token 'three': was expecting ('true', 'false' or 'null'); nested exception is com.fasterxml.jackson.core.JsonParseException: Unrecognized token 'aaa': was expecting ('true', 'false' or 'null')\n at [Source: java.io.PushbackInputStream@cba7ebc; line: 4, column: 17]",
 "timestamp": "2019-03-05T04:02:30.627Z",

 "status": 400,
 "exception": "org.springframework.http.converter.HttpMessageNotReadableException",
 "path": "/birds"
}
```

In my opinion the ideal mechanism is: (sorry for asking too much)

1. Worker can throw error code, error message, and optionally an object/dict containing variables (above example: `status`, `exception`, and `path`, and can nest objects like usual)
2. These will be merged into one error object containing `error`, `message`, `timestamp` (autogenerated), and additional variables.
3. In the BPMN file, if user wants to save the error object, user sets "Error variable" in Zeebe Modeler, e.g. `error`. The next tasks will be able to get the that error object and perform any logic, or it can also do gateway just like ordinary variables, etc.
4. `timestamp` in ISO 8601 format with millisecond precision and UTC time zone. (ISO8601 is strict and avoids ambiguity when interpreting timestamp)
5. when catching error, it's possible to use wildcard for `error` code, each segment separated by `.`. For example, the Boundary Event can catch `com.mycompany.http.*` or even `*`.

I realize above points expand the scope on what's originally requested. The most priority one in this case is supporting the error code, error message, and the custom variables, before the additional features.

**Describe alternatives you've considered**
Alternative is to use interrupting message catch boundary event, but is semantically different and behavior is different when uncaught.

**Additional context**
Related to: #3497, #3801.

### LINKED ISSUE #9473: Catch error message/details on a boundary event
**Is your feature request related to a problem? Please describe.**
Trying to capture the error message/description thrown by a worker within `ERROR BOUNDARY EVENT`.

For example I have the following workflow:
image (4)

All workers (Service 1 and 2) have same `errorCode` value defined in the Header:
image (5)

Therefore, if any exception happens, the worker will throw the error with that code and it will be handled by the Boundary Event:
image (6)

For workers, I am using the npm library with the following logic:
```
try {
  // do worker logic
  job.complete();
} catch (error) {
  const errorCode: string = job.customHeaders.errorCode;
  job.error(errorCode, error.message);
}
```
As you can see, the client library allow it to pass an optional error message as the second parameter. And that's the message I am trying to capture in the BPMN model after catching the error and using it on Service 3.

**Describe the solution you'd like**
I would like to receive that error message thrown by any Service inside that subprocess (1 or 2) as output variable and use it in the Service 3.

**Describe alternatives you've considered**
I thought one way of handling it by having multiple Error Boundary Events (each one with an unique errorCode). Then, I would be able to map those failures to know where it failed (in Service 1 or 2). However, I won't have specific details about the error (e.g. stack trace). Another problem with this approach is that I have multiple workers inside a subprocess. This is just a sample workflow. My workflows are much more complex than that and it would require many Boundary Events.

**Additional context**
I already seen #4337  and #4080 topics while investigating alternatives and I couldn't find what I am looking for.

### PR DESCRIPTION
**Pull request by skayliu**
## Description

Support throw error with variables.

image

## Related issues

closes #4337
closes #4080 
closes #9473 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by skayliu** [comment]
Hi, @saig0, @korthout. Please take a moment to review this feature when you have time. Thanks. ❤️

**Comment by korthout** [comment]
Thanks @skayliu 

I quickly noticed that your PR states:

>closes https://github.com/camunda/zeebe/issues/4337
closes https://github.com/camunda/zeebe/issues/4080
closes https://github.com/camunda/zeebe/issues/9473

Both #4337 and #9473 differ from #4080 (and your PR) in that they refer explicitly to the error message and error code properties and somehow extract them into variables, while your PR is about passing along variables in the thrown error.

🔧 Let's only close #4080.
🔧 Let's add a reference `relates to #4337`

I'll make sure to provide a full review of this PR in the coming days.

**Comment by skayliu** [comment]
Hi

File: engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobThrowErrorTest.java [hunk 1]
```diff
@@ -100,4 +113,153 @@ public void shouldRejectIfErrorIsThrown() {
...
+            tuple("foo", "\"bar\"", errorEvent.getKey(), VariableIntent.CREATED),
+            tuple("foo", "\"bar\"", processInstanceKey, VariableIntent.CREATED));
+  }
+
+  @Test
+  public void shouldThrowErrorWithExclusiveVariables() {
+    // given
+    final BpmnModelInstance modelInstance =
+        Bpmn.readModelFromStream(
+            JobThrowErrorTest.class.getResourceAsStream("/processes/error_event_exclusive.bpmn"));
```
Comment (korthout) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobThrowErrorTest.java:213]: ❓ What is this test verifying? Please clarify

File: gateway-protocol/src/main/proto/gateway.proto [hunk 2]
```diff
@@ -291,6 +291,12 @@ message ThrowErrorRequest {
   string errorCode = 2;
   // an optional error message that provides additional context
   string errorMessage = 3;
+  // JSON document that will instantiate the variables at the local scope of the
+  // job's associated task, and will propagate to upper scopes; it must be a JSON object, as variables will be mapped in a
```
Comment (korthout) [inline gateway-protocol/src/main/proto/gateway.proto:295]: >// job's associated task, and will propagate to upper scopes;

❌ The variables should only be merged on the error catch event that catches the thrown error. It should not be propagated, unless output mappings are used to achieve this.

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks for this amazing contribution @skayliu 🚀 

There are a few things we need to adjust, but you've done a wonderful job!

Please have a look at my requests. 🙇

File: engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobThrowErrorTest.java [hunk 3]
```diff
@@ -100,4 +113,153 @@ public void shouldRejectIfErrorIsThrown() {
...
+    assertThat(variableRecords)
+        .extracting(
+            r -> r.getValue().getName(),
+            r -> r.getValue().getValue(),
+            r -> r.getValue().getScopeKey(),
+            r -> r.getIntent())
+        .containsSubsequence(
+            tuple("foo", "\"bar\"", job.getValue().getElementInstanceKey(), VariableIntent.CREATED),
+            tuple("foo", "\"bar\"", errorEvent.getKey(), VariableIntent.CREATED),
+            tuple("foo", "\"bar\"", processInstanceKey, VariableIntent.CREATED));
```
[the remark this answers is not in the document]
Comment (skayliu) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobThrowErrorTest.java:205]: And what about this one, right row the variable is created at the  process instance, not at the `event sub process`, or the `error start event`.
error-event-process

[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/job/JobThrowErrorTest.java]: Thanks for investigating @skayliu. It sounds related to #9061, but it's likely a separate bug. I've started investigating this in some more detail and will follow-up on this.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/error/ErrorCatchEventTest.java [hunk 4]
```diff
@@ -209,4 +217,59 @@
...
+
+    // when
+    final Record<JobRecordValue> error =
+        ENGINE
+            .job()
+            .ofInstance(processInstanceKey)
+            .withType(JOB_TYPE)
+            .withErrorCode(ERROR_CODE)
+            .withVariables("{'foo':'bar'}")
+            .throwError();
```
[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/error/ErrorCatchEventTest.java:236]: @skayliu Please remove the variable assignment.
```suggestion
    ENGINE
        .job()
        .ofInstance(processInstanceKey)
        .withType(JOB_TYPE)
        .withErrorCode(ERROR_CODE)
        .withVariables("{'foo':'bar'}")
        .throwError();
```

**Review by korthout — CHANGES_REQUESTED** [review]
Thanks @skayliu I have one last request and then this can be merged 🥇 

👍 About the variable scopes, it seems that I was wrong. The way you've implemented/tested it is correct. I'll discuss it once more with the team next week, but that doesn't have to block this PR.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnVariableMappingBehavior.java [hunk 5]
```diff
@@ -148,4 +149,13 @@ private boolean isConnectedToEventBasedGateway(final ExecutableFlowNode element)
...
   }
+
+  private boolean isConnectedToErrorEvent(final ExecutableFlowNode element) {
+    if (element instanceof ExecutableCatchEventElement) {
+      final var catchEvent = (ExecutableCatchEventElement) element;
+      return catchEvent.isError();
+    } else {
+      return false;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (skayliu) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/behavior/BpmnVariableMappingBehavior.java:160]: what name should be?

**Comment by skayliu** [comment]
Hi, @korthout , It's done, please check this out. Thanks.

**Comment by korthout** [comment]
@oleschoenburg Recently you made [changes to the caching](https://github.com/camunda/zeebe/pull/10662). The test [run](https://github.com/camunda/zeebe/actions/runs/3445383381/jobs/5750593349) on this PR fails because after the Zeebe Gateway has been built, the used cache is replaced by another cache. Do you agree that we should disable the cache [when switching the java version to 8](https://github.com/camunda/zeebe/actions/runs/3445383381/workflow#L292)?

- Cache before building: `Linux-maven-java-client-e9dbd284a3699116d5717077527466fe949957b550bcb2bbd77278ecafe4a747`
- Cache after builder, before starting the test: `setup-java-Linux-maven-e9dbd284a3699116d5717077527466fe949957b550bcb2bbd77278ecafe4a747`

Screen Shot 2022-11-11 at 17 10 23

@skayliu Caching is turned off by default, so I believe we can try to fix this by removing [this line](https://github.com/skayliu/zeebe/blob/4337-set-variables-for-throw-error/.github/workflows/ci.yml#L292). Can you please remove that line in another commit to this PR?

**Comment by skayliu** [comment]
> Caching is turned off by default, so I believe we can try to fix this by removing [this line](https://github.com/skayliu/zeebe/blob/4337-set-variables-for-throw-error/.github/workflows/ci.yml#L292). Can you please remove that line in another commit to this PR?

Hi, @korthout, It's been removed. Please try again.

**Comment by korthout** [comment]
@skayliu Thanks for the quick change. It worked! 💯 

🔧 Please also squash your commits into ones with more meaningful descriptions. Both our [Contributing Guide](https://github.com/camunda/zeebe/blob/main/CONTRIBUTING.md#commit-message-guidelines) and our Wiki entry on [Pull Requests](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#ive-written-a-clear-changelist-description) push for more explanation in commit messages. This can help us understand why certain changes are made during the review. But it can also help us when we're trying to understand code in the future because the choices are written close the code (in the version control). It will help us if you try to provide a bit more info in the commit messages about the changes. 🙇  Thanks a lot

**Comment by lzgabel** [comment]
Done.

Hi @korthout. Please check this out. 🙇

**Comment by oleschoenburg** [comment]
@korthout Yes, sounds reasonable :+1: No cache needed for Java 8, Zeebe should be built already, we are just switching to Java 8 to run tests. Nice catch!

**Review by korthout — APPROVED** [review]
Thanks @skayliu 🚀 

Let's get this merged!

bors merge

PS: Sorry for the delay in my review 🙇

**Comment by korthout** [comment]
It seems Bors got stuck

>Resource not accessible by integration

It seems to be related to the changes to the workflow. Bors is not allowed to merge PRs from forks with workflow changes due to permissions.

@skayliu Do you mind if I extract that commit into another PR?

**Comment by skayliu** [comment]
> @skayliu Do you mind if I extract that commit into another PR?

@korthout, never mind, just go ahead.

**Comment by korthout** [comment]
@skayliu https://github.com/camunda/zeebe/pull/10995 has been merged. Please remove the commit [695bc95](https://github.com/camunda/zeebe/pull/10882/commits/695bc954533b645c1152cad6e8d3a1aa692dea3f) and rebase on top of `main`. Then we should be able to get this merged.

**Comment by lzgabel** [comment]
Done.

Hi @korthout. Please check this out. 🙇

**Review by korthout — APPROVED** [review]
LGTM 🚀 

bors merge

**Comment by korthout** [comment]
bors cancel

**Comment by lzgabel** [comment]
Sorry.  @korthout. Please try it again. 🙇

**Comment by korthout** [comment]
@skayliu Seems there's a failing test now.

```
JobThrowErrorTest.shouldThrowErrorWithVariablesWithEventSubProcess:388 [With event sub process the variables are created at the process instance] 
Expecting actual:
  [("foo", ""bar"", 2251799813685273L, CREATED),
    ("foo", ""bar"", 2251799813685269L, CREATED)]
to contain exactly (and in same order):
  [("foo", ""bar"", 2251799813685269L, CREATED)]
but some elements were not expected:
  [("foo", ""bar"", 2251799813685273L, CREATED)]
```

**Comment by lzgabel** [comment]
Sorry @korthout. :bow: It was caused by the conflict resolution during the rebase on `main`, and I've resolved it. Wait a few minutes. 🙇

**Comment by korthout** [comment]
Looks good to me! Thanks @lzgabel and @skayliu 

bors merge

**Comment by skayliu** [comment]
Hi, @korthout The Go client is not include, Please open an issue for it.

**Comment by korthout** [comment]
Thanks for noting that @skayliu, I've opened an issue:
- #11019

**Pull request merged**